lib/Quantization/Quantization.cpp

```diff
@@ -502,31 +507,68 @@ class FunctionQuantizer : public FunctionConverter {
-      if (auto *Q = llvm::dyn_cast<DequantizeNode>(&node)) {
-        if (auto *fcN = llvm::dyn_cast<FullyConnectedNode>(Q->getInput())) {
-          NodeValue input = fcN->getInput();
-          NodeValue weights = fcN->getWeights();
-          NodeValue bias = fcN->getBias();
-          NodeValue result = fcN->getResult();
-          // Only convert quantized FullyConnected Node.
-          if (input.getType()->isQuantizedType() &&
-              llvm::isa<QuantizeNode>(weights.getNode()) &&
-              bias.getType()->isQuantizedType() &&
-              result.getType()->isQuantizedType()) {
-            auto *wq = llvm::dyn_cast<QuantizeNode>(weights.getNode());
-            // For RowwiseQuantizedFullyConnected, the weights need to be
-            // constant.
-            if (Constant *wc = llvm::dyn_cast<Constant>(wq->getInput())) {
-              auto *fcq = function_.createRowwiseQuantizedFullyConnected(
-                  "rowwiseqfc", input, wc, bias, result.getType(),
-                  /* transposeWeight */ true);
-              // Replace the usages of quantized FC node to
-              // RowwiseQuantizedFullyConnected Node.
-              result.replaceAllUsesOfWith(fcq->getResult());
-            }
+      if (auto *fcN = llvm::dyn_cast<FullyConnectedNode>(Q->getInput())) {
+        NodeValue input = fcN->getInput();
+        NodeValue weights = fcN->getWeights();
+        NodeValue bias = fcN->getBias();
+        NodeValue result = fcN->getResult();
+        // Only convert quantized FullyConnected Node.
+        if (input.getType()->isQuantizedType() &&
+            llvm::isa<QuantizeNode>(weights.getNode()) &&
+            bias.getType()->isQuantizedType() &&
+            result.getType()->isQuantizedType()) {
+          auto *wq = llvm::dyn_cast<QuantizeNode>(weights.getNode());
+          // For RowwiseQuantizedFullyConnected, the weights need to be
+          // constant.
+          if (Constant *wc = llvm::dyn_cast<Constant>(wq->getInput())) {
+            auto *fcq = function_.createRowwiseQuantizedFullyConnected(
+                "rowwiseqfc", input, wc, bias, result.getType(),
+                /* transposeWeight */ true);
+            // Replace the usages of quantized FC node to
+            // RowwiseQuantizedFullyConnected Node.
+            result.replaceAllUsesOfWith(fcq->getResult());
           }
         }
       }
+
+      // Convert SLWS from normal quantized version to fused rowwise-quantized
+      // version if applicable. Data must be Constant for this to occur. We also
+      // will not quantized the weights as we do for the default normal
+      // quantized SLWS, as the rowwise version uses float weights.
+      if (auto *SLWS =
+              llvm::dyn_cast<SparseLengthsWeightedSumNode>(Q->getInput())) {
+        NodeValue data = SLWS->getData();
+        auto *dataQN = llvm::dyn_cast<QuantizeNode>(data.getNode());
+        assert(dataQN && "Data input should have been quantized");
```

Comment:
any reason those checks are asserts vs for FC it's actually part of the precondition before conversion?

Reply:
Hmm, well I guess it's possible that the FC was not quantized due to `doNotQuantizeKinds`. I'll make it a precondition, thanks!